### PR TITLE
Story/cite 171

### PR DESCRIPTION
**Pull request by PratikGiri**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

Added a Person Index page for groups.
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+@Controller
+public class ShowPeopleController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+    
+    Map<IPerson, List<ICitation>> iPersoniCitSortedMap = new TreeMap<IPerson, List<ICitation>>();
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:40]: Use appropriate variable names. Makes it easier to read.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/showPeople")
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:42]: Path variables should be part of the end-point

**Review by MaulikLim — COMMENTED** [review]
I'm guessing that the indentation and code factor issues will be resolved when PR is ready to be reviewed

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+@Controller
+public class ShowPeopleController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+
+    Map<IPerson, List<ICitation>> personCitationSortedMap = new TreeMap<IPerson, List<ICitation>>();
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:40]: should be private

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/showPeople")
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:42]: should be `.../items/people`

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        Map<IPerson, List<ICitation>> personCitationMap = new HashMap<IPerson, List<ICitation>>();
+
+        results.getCitations().stream()
+                .forEach(
+                        iCitation -> (Stream
+                                .of(iCitation.getAuthors().stream(), iCitation.getEditors().stream(),
+                                        iCitation.getOtherCreators().stream()
+                                                .map(iOtherCreator -> iOtherCreator.getPerson()))
+                                .reduce(Stream::concat).orElseGet(Stream::empty))
+                                        .filter(distinctByKey(p -> p.getFirstName())).forEach(au -> {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:76]: what if there are two people with the same first name?

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                                        }));
+
+        Map<IPerson, List<ICitation>> personCitationTMap = new TreeMap<IPerson, List<ICitation>>(personCitationMap);
+
+        Collection<IPerson> keyCollection = personCitationTMap.keySet();
+        List<IPerson> peopleList = new ArrayList<IPerson>(keyCollection);
+
+        model.addAttribute("people", peopleList);
+
+        personCitationSortedMap = new TreeMap<IPerson, List<ICitation>>(personCitationTMap);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:101]: all of the above is business logic and should not be in the controller

Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:101]: also, spring beans are singletons by default. If someone else goes to a person index of another group, this will be overwritten with their map.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                                                    break;
+                                                }
+                                            }
+                                            if (!containKey) {
+                                                List<ICitation> citList = new ArrayList<ICitation>();
+                                                citList.add(iCitation);
+                                                personCitationMap.put(au, citList);
+                                            }
+
+                                        }));
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:92]: all the different nested streams here make it very hard to follow. This is a case were mixing streams and regular loops would be preferable.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        Integer pageInt = 1;
+        try {
+            pageInt = new Integer(page);
+        } catch (NumberFormatException ex) {
+            logger.warn("Trying to access invalid page number: " + page);
+        }
+        IUser user = (IUser) authentication.getPrincipal();
+        CitationResults results;
+        try {
+            results = citationManager.getGroupItems(user, groupId, null, pageInt, sort);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ShowPeopleController.java:55]: this only get the citations on one page, doesn't it? the person index should be for the whole collection.

I would try to do as much as you can through aggregations in the db, otherwise we'll easily run into memory issues (some groups have > 20,000 entries).

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/showCitations.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+<body>
+  <div layout:fragment="content">
+    <ul class="list-group">
+      <li th:each="iCitation, iterStat : ${involvedCitations}" class="list-group-item clearfix">
+        <a class="bib-entry"
+          th:href="@{|/auth/group/${zoteroGroupId}/items/${iCitation.key}?index=${iterStat.index}|}">
+          <i>[[${iCitation.title}]] </i>
+        </a>
+      </li>
+    </ul>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/showCitations.html:16]: this list needs to have all the information that the items list page has

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/showPeople.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  <script src="https://code.jquery.com/ui/1.10.1/jquery-ui.min.js"></script>
+</head>
+
+<body>
+  <div layout:fragment="content">
+    <ul class="list-group">
+      <li th:each="person, iterStat : ${people}" class="list-group-item clearfix">
+
+        <a th:href="@{|/auth/group/${zoteroGroupId}/items/showCitation/${iterStat.index}|}">
+          <i>[[${person.firstName}]] [[${person.lastName}]]</i>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/showPeople.html:13]: there should also be the uri here.

**Review by jdamerow — CHANGES_REQUESTED** [review]
